pkg/util/exec/exec_pod.go

```diff
@@ -0,0 +1,134 @@
+limitations under the License.
+*/
+
+// Heavily inspired by https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/exec_util.go
+
+package exec
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/remotecommand"
+)
+
+var (
+	CephCommandTimeout = 15 * time.Second
+)
+
+// ExecOptions passed to ExecWithOptions
+type ExecOptions struct {
+	Command       []string
+	Namespace     string
+	PodName       string
+	ContainerName string
+	Stdin         io.Reader
+	CaptureStdout bool
+	CaptureStderr bool
+	// If false, whitespace in std{err,out} will be removed.
+	PreserveWhitespace bool
+}
+
+// CommandExecutorExecInPod is an exec.Executor that translates every command before executing it
+// This is useful to run the commands in a job with `kubectl run ...` when running the operator outside
+// of Kubernetes and need to run tools that require running inside the cluster.
+type CommandExecutorExecInPod struct {
+	ClientSet  kubernetes.Interface
+	RestClient *rest.Config
+	Namespace  string
+}
+
+// ExecWithOptions executes a command in the specified container,
+// returning stdout, stderr and error. `options` allowed for
+// additional parameters to be passed.
+func (e *CommandExecutorExecInPod) ExecWithOptions(options ExecOptions) (string, string, error) {
+	const tty = false
+
+	logger.Debugf("ExecWithOptions %+v", options)
+
+	req := e.ClientSet.CoreV1().RESTClient().Post().
+		Resource("pods").
+		Name(options.PodName).
+		Namespace(options.Namespace).
+		SubResource("exec").
+		Param("container", options.ContainerName)
+	req.VersionedParams(&v1.PodExecOptions{
+		Container: options.ContainerName,
+		Command:   options.Command,
+		Stdin:     options.Stdin != nil,
+		Stdout:    options.CaptureStdout,
+		Stderr:    options.CaptureStderr,
+		TTY:       tty,
+	}, scheme.ParameterCodec)
+
+	var stdout, stderr bytes.Buffer
+	err := execute("POST", req.URL(), e.RestClient, options.Stdin, &stdout, &stderr, tty)
+
+	if options.PreserveWhitespace {
+		return stdout.String(), stderr.String(), err
+	}
+	return strings.TrimSpace(stdout.String()), strings.TrimSpace(stderr.String()), err
+}
+
+// ExecCommandInContainerWithFullOutput executes a command in the
+// specified container and return stdout, stderr and error
+func (e *CommandExecutorExecInPod) ExecCommandInContainerWithFullOutput(appLabel, containerName string, cmd ...string) (string, string, error) {
+	options := metav1.ListOptions{LabelSelector: fmt.Sprintf("app=%s", appLabel)}
+	pods, err := e.ClientSet.CoreV1().Pods(e.Namespace).List(context.TODO(), options)
+	if err != nil {
+		return "", "", err
+	}
+
+	return e.ExecWithOptions(ExecOptions{
+		Command:            cmd,
+		Namespace:          e.Namespace,
+		PodName:            pods.Items[0].Name,
```

Comment:
Do we need to handle the case where `pods.Items` is zero length?

Reply:
Future thought: if there are more than one items, we could try each result if the command fails to run due to a network partition-related error.

Reply:
Yes I as thinking about the stretched scenario as well, will see what I can do in this PR.